README.md
[![GitHub license](https://img.shields.io/github/license/Rollcloud/whathappened)](https://github.com/Rollcloud/whathappened/blob/main/LICENSE)
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/Rollcloud/whathappened/test-build?logo=github)](https://github.com/Rollcloud/whathappened/actions?query=workflow%3Atest-build)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/whathappened?logo=pypi)](https://pypi.org/project/whathappened/)
[![semver](https://img.shields.io/badge/semver-2.0.0-blue)](https://semver.org/)
[![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/rollcloud/whathappened?sort=semver)](https://github.com/Rollcloud/whathappened/releases)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

# whathappened
A changelog generator using simply structured git commit messages

## Inspired by

* [SemVer](https://semver.org/)
* [Angular Commit Message Format](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit)
* [Auto Changelog](https://github.com/Michael-F-Bryan/auto-changelog)
* [git_commits.py](https://gist.github.com/simonw/091b765a071d1558464371042db3b959#file-get_commits-py)

## Install and Run

Installation is as simple as it gets:

    $ pip install whathappened

To generate a changelog, run:

    $ whathappened

Make sure to activate any virtual envrionment that you might be using first.

Some options are available:

    $ whathappened --help
    Usage: whathappened [OPTIONS] [GIT_LOG_ARGS]...

      Handle command line arguments. Extra arguments are passed to 'git log'.

    Options:
      --overriding-version TEXT  Specify a version number to use [format: x.y.z]
      -o, --output PATH  The changelog destination [default: stdout]
      -e, --emoji        Include emoji in headings if present
      -p, --prefix TEXT  Version prefix, often 'version' or 'v' [default: '']
      
      --version          Show the version and exit.
      --help             Show this message and exit.

To limit the range of commits, add a [`revision range`](https://git-scm.com/docs/git-log#Documentation/git-log.txt-ltrevisionrangegt) argument.

To create a changelog of commits that are yet to be pushed:

    $ whathappened origin..HEAD

To create a changelog of commits between (branch: main) and (tag: v1.0.0) and write the output to file:

    $ whathappened -o CHANGELOG.md main..v1.0.0

## Package Changelog

Created by Whathappened itself - very meta.

For the complete changelog, please see [CHANGELOG.md](CHANGELOG.md).

## Whathappened Commit Message Format

Whathappened expects git commit messages in the format outlined below:

    [optional breaking ]<type>[ optional (<scope>)]: <description>

    [optional body]

`<type>` is recommended to be one of:

    fix
    feat
    build
    ci
    docs
    style
    refactor
    perf
    test

Variations on these types like `feat, feature, features`, or `doc, docs` are seamlessly grouped together.

`<scope>` is recommended to be a module, file, or folder name as appropiate.

For a nice summary of `type`s and what they mean, please see [Understanding Semantic Commit Messages Using Git and Angular](https://nitayneeman.com/posts/understanding-semantic-commit-messages-using-git-and-angular/).

## Emoji

Adding the --emoji flag will add emoji after each heading in the changelog. Below is the list of emoji that are used:

* Docs 📝
* Features ✨
* Fixes 🐛
* Performance ⚡️
* Refactorings ♻️
* Other 🃏 

### Examples

All of the examples below are acceptable commit message formats:

The simplest message:

    style: isn't it neat

Defining scope:

    docs(gadgets): remember the gizmos

Signalling a breaking change

    breaking feat: add the whozits

Defining scope, including a space:

    fix (whatsits): repair the thingamabobs

Multiple lines:

    test: How many have I got?

    I've got twenty
    But who cares?

    I want more

For more examples, please see whathappened's own [commits](https://github.com/Rollcloud/whathappened/commits/).


CHANGELOG.md
# Changelog

## v1.0.0 (2021-04-23)

### Docs

* Changelog - update to v1.0.0
* Readme - correct minor spelling errors
* Readme - update examples to point to Commit-It-Simple website

### Features

* Changelog - allow the latest version number to be overridden

### Fixes

* Example - add Python3 shebang


## v0.3.4 (2020-11-28)

### Docs

* Changelog - updated to v0.3.4

### Features

* Changelog - implement emoji headings

### Fixes

* Changelog - do not display Other if it is the only version heading
* Changelog - escape underscores for correct Markdown rendering

### Refactorings

* Changelog - rename internal function sentence() -> \_sentence()


## v0.3.3 (2020-11-12)

### Docs

* Changelog - updated to v0.3.3
* Readme - add command-line version prefix specifier

### Fixes

* Changelog - improve robustness of category grouping


## v0.3.2 (2020-11-08)

### Docs

* Changelog - updated to v0.3.2

### Features

* Changelog - count breaking, feature, & fix commits in each version
* Changelog - replace "HEAD" with the predicted Semver version
* Command-line - allow a version prefix to be specified

### Refactorings

* Make get\_commits() accessible from changelog module


## v0.3.1 (2020-10-19)

### Docs

* Changelog - updated to v0.3.1

### Fixes

* Ci - ignore erroneous F821 error when fetching package version


## v0.3.0 (2020-10-19)

### Docs

* Changelog - update to v0.3.0

### Features

* Add basic command line options [BREAKING]

### Fixes

* Gracefully handle commits like 'filename.py: message'
* Increase robustness of scope by allowing spaces, dots, etc.


## v0.2.0 (2020-10-18)

### Docs

* Changelog - update for v0.2.0
* Readme - add command-line indicator to installation instructions
* Readme - add commit message examples
* Readme - add more details about message types
* Readme - link badges to sensible destinations

### Features

* Add command line script
* Detect and handle breaking changes
* Group commits by version [BREAKING]
* Group commits in each version by type

### Fixes

* Add appropriate capitalisation to changelog items
* Only display user-relevant commits types in the changelog
* Remove extra line at end of changelog


## v0.1.1 (2020-10-18)

### Docs

* Changelog - update for v0.1.1
* Readme - add Python version badge
* Readme - add installation, execution, and changelog instructions
* Readme - move broken 'commits since' badge


## v0.1.0 (2020-10-17)

### Docs

* Changelog - add initial CHANGELOG.md
* Git_commits - add credit to original source
* Readme - add badges
* Readme - add inspiration
* Readme - specify expected message format

### Features

* Add changelog generation script
* Add git\_commits.py from existing Gist
* Create example.py

### Other

* Initial commit

### Refactorings

* Git_commits - convert to Python3


tests/test_git_commits.py
import pytest

from whathappened.git_commits import get_commits


with open('tests/git-log.out') as f:
    test_log = f.read().encode("utf-8")


def mock_check_output(args, **kwargs):
    return test_log


def test_get_commits(monkeypatch):
    monkeypatch.setattr('subprocess.check_output', mock_check_output)

    commits = get_commits()

    print(commits)

    expected = [
        {
            'hash': 'e324c324df48a76113ad9b3c0887f161324046e4',
            'tags': ['v0.1.1'],
            'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
            'date': 'Sat Oct 17 17:30:25 2020 +0200',
            'message': '',
            'title': 'breaking feat(readme): specify expected message format',
        },
        {
            'hash': '7b4e7e657f9e3f2f4033cc5f47bcc637f5799fe9',
            'tags': [],
            'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
            'date': 'Sat Oct 17 15:00:48 2020 +0200',
            'message': '',
            'title': 'breaking fix (code): repair things',
        },
        {
            'hash': '7b4e7e657f9e3f2f4033cc5f47bcc637f5799fe9',
            'tags': [],
            'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
            'date': 'Sat Oct 17 15:00:48 2020 +0200',
            'message': '',
            'title': 'breaking fix: repair things',
        },
        {
            'hash': '7b4e7e657f9e3f2f4033cc5f47bcc637f5799fe9',
            'tags': [],
            'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
            'date': 'Sat Oct 17 15:00:48 2020 +0200',
            'message': '',
            'title': 'fix: add inspiration',
        },
        {
            'hash': 'f60445bba0ac48e12ce6be5526644037234ae500',
            'tags': ['v0.0.1'],
            'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
            'date': 'Sat Oct 17 15:00:31 2020 +0200',
            'message': '',
            'title': 'docs (readme): add badges',
        },
        {
            'hash': '9e57ba91f54244af913931c017480a39605c15f9',
            'tags': [],
            'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
            'date': 'Sat Oct 17 13:55:04 2020 +0200',
            'message': (
                'Setup Python 3.6 on ubuntu-latest\n'
                'Install pipenv and dependencies\n'
                'Test'
            ),
            'title': 'build(actions): create python-app.yml for github actions',
        },
        {
            'hash': '4094d22846daea951c4fe0d74abb2a798e9a3404',
            'tags': ['v0.0.0'],
            'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
            'date': 'Sat Oct 17 13:19:28 2020 +0200',
            'message': '',
            'title': 'Initial commit',
        },
    ]

    assert commits == expected


@pytest.mark.parametrize(
    "test_input, expected",
    [
        (
            (
                'commit 7b4e7e657f9e3f2f4033cc5f47bcc637f5799fe9\n'
                'Author: Rollcloud <Rollcloud@users.noreply.github.com>\n'
                'Date:   Sat Oct 17 15:00:48 2020 +0200\n'
                '\n'
                '    fix: check _underscore_ symbol\n'
            ),
            [
                {
                    'hash': '7b4e7e657f9e3f2f4033cc5f47bcc637f5799fe9',
                    'tags': [],
                    'author': 'Rollcloud <Rollcloud@users.noreply.github.com>',
                    'date': 'Sat Oct 17 15:00:48 2020 +0200',
                    'message': '',
                    'title': 'fix: check _underscore_ symbol',
                }
            ],
        )
    ],
)
def test_get_commits_symbols(test_input, expected, monkeypatch):
    monkeypatch.setattr(
        'subprocess.check_output', lambda *args, **kwargs: test_input.encode("utf-8")
    )

    commits = get_commits()
    print(commits)

    assert commits == expected


tests/misc/pytest-summary.py
import json
import sys


def load_report(filename):
    report = []
    with open(filename, "r") as f:
        for line in f:
            result = json.loads(line)
            if result["$report_type"] == "TestReport" and result["when"] == "call":
                report.append(result)
            if (
                result["$report_type"] == "CollectReport"
                and result["outcome"] == "failed"
            ):
                result["outcome"] = "errors"  # pytest report shows error
                report.append(result)

    return report


def print_report(report):
    summary = {}
    for row in report:
        if row["outcome"] not in summary:
            summary[row["outcome"]] = 0
        summary[row["outcome"]] += 1

    summary_text = ', '.join([f"{v} {k}" for k, v in summary.items()])

    print("Results: " + summary_text)


def main():
    filename = sys.argv[1]
    report = load_report(filename)
    print_report(report)


if __name__ == '__main__':
    main()


whathappened/git_commits.py
# https://gist.github.com/simonw/091b765a071d1558464371042db3b959#file-get_commits-py

import subprocess
import re

leading_4_spaces = re.compile('^    ')


def get_commits():
    lines = (
        subprocess.check_output(['git', 'log'], stderr=subprocess.STDOUT)
        .decode("utf-8")
        .split('\n')
    )
    commits = []
    current_commit = {}

    def save_current_commit():
        title = current_commit['message'][0]
        message = current_commit['message'][1:]
        if message and message[0] == '':
            del message[0]
        current_commit['title'] = title
        current_commit['message'] = '\n'.join(message)
        commits.append(current_commit)

    for line in lines:
        if not line.startswith(' '):
            if line.startswith('commit '):
                if current_commit:
                    save_current_commit()
                    current_commit = {}
                current_commit['hash'] = line.split('commit ')[1]
            else:
                try:
                    key, value = line.split(':', 1)
                    current_commit[key.lower()] = value.strip()
                except ValueError:
                    pass
        else:
            current_commit.setdefault(
                'message', []
            ).append(leading_4_spaces.sub('', line))
    if current_commit:
        save_current_commit()
    return commits


whathappened/command_line.py
import click

try:
    from whathappened import __version__, changelog
except ImportError:  # for development use
    import changelog

    __version__ = 'major.minor.patch'


def main(output="CHANGELOG.md", emoji=False, git_log_args=[]):
    commits = changelog.get_commits(git_log_args=git_log_args)
    versions = changelog.compile_log(commits)
    versions = changelog.update_latest_version(versions)
    log = changelog.format_log(versions, emoji=emoji)

    if output is not None:
        changelog.write_log(log, output)
    else:
        click.echo(log)


@click.command(context_settings=dict(ignore_unknown_options=True))
@click.option(
    '--output',
    '-o',
    type=click.Path(),
    default=None,
    help="The changelog destination [default: stdout]",
)
@click.option(
    '--emoji',
    '-e',
    is_flag=True,
    default=False,
    help="Include emoji in headings if present",
)
@click.argument('git_log_args', nargs=-1, type=click.UNPROCESSED)
@click.version_option(version=__version__)
def cli(output, emoji, git_log_args):
    """
    Handle command line arguments. Extra arguments are passed to 'git log'.
    """
    main(output=output, emoji=emoji, git_log_args=git_log_args)


if __name__ == '__main__':
    cli()


whathappened/changelog.py
import re

from datetime import datetime
from itertools import groupby

# imported to make this module the base module for all external calls to functions
try:
    from whathappened.git_commits import get_commits  # noqa
except ImportError:  # for development use
    from git_commits import get_commits  # noqa

# https://semver.org/spec/v2.0.0.html
semver_regex = re.compile(
    r"^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)"
    r"(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)"
    r"(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
    r"(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"
)


class VersionFormatException(Exception):
    """Raise when a version string cannot be correctly parsed."""

    pass


class Version:
    """
    All the changes committed prior to a new version being released.

    Also includes their attributes, such as breaking, feature, or fix.
    """

    def __init__(self, ref, date):
        self.ref = ref
        self.date = (
            datetime.strptime(date, "%a %b %d %H:%M:%S %Y %z")
            if isinstance(date, str)
            else date
        )  # str input format Wed Apr 22 18:58:54 2020 +0200

        self.breaking = 0
        self.feature = 0
        self.fix = 0

        self.commits = []

    def __repr__(self):
        return (
            f"Version('{self.ref}',"
            f" {self.date},"
            f" breaking={self.breaking},"
            f" feature={self.feature},"
            f" fix={self.fix},"
            f" num_commits={len(self.commits)})"
        )  # pragma: no cover


class Commit:
    """The content and attributes of a single Git commit."""

    commit_regex = re.compile(
        r"(?:(?P<breaking>break(?:ing)?)? ?(?P<type>\w+){1}"
        r" ?(?:\((?P<scope>[^\(\):]+)\))?: (?P<description>.+)|"
        r"(?P<description_alt>.+))",
        flags=re.IGNORECASE,
    )

    def __init__(self, commit_dict):
        self.commit_dict = commit_dict

    def __getattr__(self, name):

        try:
            return self.commit_dict[name]
        except KeyError:

            result = Commit.commit_regex.match(self.title)
            breaking, commit_type, scope, description, description_alt = (
                result.group('breaking'),
                result.group('type'),
                result.group('scope'),
                result.group('description'),
                result.group('description_alt'),
            )

            if name == 'description':
                return description if description is not None else description_alt
            elif name == 'is_breaking':
                return breaking is not None
            elif name == 'type':
                # group equivalent commit types
                types = {
                    'doc': 'docs',
                    'docs': 'docs',
                    'feat': 'feat',
                    'feature': 'feat',
                    'features': 'feat',
                    'fix': 'fix',
                    'fixes': 'fix',
                    'perf': 'perf',
                    'performance': 'perf',
                    'refac': 'refactor',
                    'refactor': 'refactor',
                    'test': 'test',
                    'revert' : 'revert',
                    'ci': 'ci',
                    'chore': 'other',
                }
                try:
                    return types[commit_type]
                except KeyError:
                    return commit_type if commit_type is not None else 'other'
            elif name == 'scope':
                return scope
            elif name == 'is_feature':
                return "feat" in self.type.lower()
            elif name == 'is_fix':
                return "fix" in self.type.lower()
            else:
                raise AttributeError(f"Attribute '{name}' not found in class Commit")

    def __repr__(self):
        return (
            f"Commit({{"
            f"'hash': '{self.hash[:6]}', "
            f"'title': '{self.title}', "
            f"}})"
        )  # pragma: no cover


def _sentence(string):
    """Format a given string in sentence case."""
    try:
        return string[0].upper() + string[1:]
    except IndexError:
        # zero-length string
        return string


def calculate_next(versions, prefix=""):
    """
    Calculate the next version number to be released.

    Based on changes made since the previous version.
    """
    global semver_regex

    try:
        previous = versions[1]
    except IndexError:
        # if no previous version has been found
        return f"{prefix}0.1.0"

    previous_version = previous.ref[len(prefix) :]
    result = semver_regex.match(previous_version)

    # extract version numbers if possible:
    try:
        major, minor, patch = (
            int(result.group('major')),
            int(result.group('minor')),
            int(result.group('patch')),
        )
    except AttributeError:
        raise VersionFormatException(
            f"The version number of '{previous_version}' with prefix='{prefix}' cannot"
            f" be parsed. Please enter the appropriate prefix or use a version string"
            f" like 'X.Y.Z'. See https://semver.org/spec/v2.0.0.html for more details."
        )

    latest_version = versions[0]

    if major == 0:
        # this is a development release and only receives minor and patch increments
        if latest_version.breaking > 0:
            return f"{prefix}{major}.{minor+1}.0"
        elif latest_version.feature > 0 or latest_version.fix > 0:
            return f"{prefix}{major}.{minor}.{patch+1}"
        else:
            # no api changes
            return f"{prefix}{major}.{minor}.{patch}"

    else:
        # this is production release and receives major, minor, and patch increments
        if latest_version.breaking > 0:
            return f"{prefix}{major+1}.0.0"
        elif latest_version.feature > 0:
            return f"{prefix}{major}.{minor+1}.0"
        elif latest_version.fix > 0:
            return f"{prefix}{major}.{minor}.{patch+1}"
        else:
            # no api changes
            return f"{prefix}{major}.{minor}.{patch}"


def compile_log(commits):
    """Iterate though a list of Commits, compiling a list of Versions based on tags."""
    versions = []

    # iterate through commits from latest to earliest

    # group by version
    for commit in commits:
        # make a new version if required
        if len(commit['tags']) > 0 or len(versions) == 0:
            tag = commit['tags'][0] if len(commit['tags']) > 0 else 'HEAD'
            versions.append(Version(ref=tag, date=commit['date']))

        this_commit = Commit(commit)

        # append to current version
        versions[-1].commits.append(this_commit)

        # check if commit is breaking, feature, or fix
        if this_commit.is_breaking:
            versions[-1].breaking += 1

        if this_commit.is_feature:
            versions[-1].feature += 1

        if this_commit.is_fix:
            versions[-1].fix += 1

    # for version in versions:
    #     print(version)

    return versions


def update_latest_version(versions, prefix=""):
    """Update the HEAD reference to show the next semver version."""
    latest_version = versions[0]
    latest_version.ref = calculate_next(versions, prefix=prefix)

    return versions


def override_latest_version(versions, overriding, prefix=""):
    """Update the HEAD reference to show the overriding version."""
    latest_version = versions[0]
    latest_version.ref = f"{prefix}{overriding}"

    return versions


def format_log(versions, emoji=False):
    """Produce a nicely formatted changelog - with emoji too if required."""
    output = "# Changelog"

    headings = {
        'docs': "Docs 📝",
        'feat': "Features ✨",
        'fix': "Fixes 🐛",
        'perf': "Performance ⚡️",
        'refactor': "Refactorings ♻️",
        'revert': "Reverted ☠️",
        'test' : "Testing 🧪",
        'ci'   : "Continous Integration 🤖",
        'other': "Other 🃏",

    }

    if not emoji:
        # remove emoji (first two characters) from headings, and strip space character
        headings = {key: heading[:-2].strip() for key, heading in headings.items()}
    else:
        # leave emoji in headings
        pass

    for version in versions:
        output += f"\n\n## {version.ref} ({version.date.isoformat()[:10]})\n"

        # store groupby results as lists
        groups = []
        uniquekeys = []
        data = sorted(version.commits, key=lambda x: x.type[:4])
        for k, g in groupby(data, lambda x: x.type):
            groups.append(list(g))  # Store group iterator as a list
            uniquekeys.append(k)

        headings_in_this_version = [k for k in uniquekeys if k in headings]

        for key, group in zip(uniquekeys, groups):
            if key in headings:

                # check if Other is the only heading in this version
                if key == 'other' and len(headings_in_this_version) == 1:
                    # if it is, it is redundent and should not be displayed
                    output += "\n"
                else:
                    # else display the heading as usual
                    output += f"\n### {headings[key]}\n\n"

                for commit in sorted(group, key=lambda x: f"{x.scope} {x.description}"):
                    scope = f"{commit.scope} - " if commit.scope else ''
                    desc = commit.description.replace('_', '\\_')  # escape underscore
                    desc = _sentence(desc) if len(scope) == 0 else desc
                    breaking = " [BREAKING]" if commit.is_breaking else ''
                    output += f"* {_sentence(scope)}{desc}{breaking}\n"

    return output


def write_log(log, filename):
    with open(filename, 'w') as f:
        f.write(log)
